File: LuaProject/Lib/Vector2.lua
local setmetatable = setmetatable
local type = type
local rawget = rawget
local m_sqrt = math.sqrt

local GetUnitX = GetUnitX
local GetUnitY = GetUnitY

---@class Vector2
local cls = {}
Vector2 = cls

cls._loc = Location(0, 0)

---@return Vector2
function cls.new(x, y)
    return setmetatable({
        x = x or 0,
        y = y or 0,
    }, cls)
end

local new = cls.new

---@param unit unit
function cls.FromUnit(unit)
    return new(GetUnitX(unit), GetUnitY(unit))
end

function cls.InsideUnitCircle()
    local angle = math.random() * math.pi * 2
    return new(math.cos(angle), math.sin(angle))
end

function cls.Dot(a, b)
    return a.x * b.x + a.y * b.y
end

function cls.Cross(a, b)
    return a.y * b.x - a.x * b.y
end

function cls.UnitDistance(u1, u2)
    local v1 = cls.FromUnit(u1)
    local v2 = cls.FromUnit(u2)
    return v1:Sub(v2):Magnitude()
end

function cls.UnitDistanceSqr(u1, u2)
    local v1 = cls.FromUnit(u1)
    local v2 = cls.FromUnit(u2)
    return v1:Sub(v2):MagnitudeSqr()
end

---@param unit unit
function cls:MoveToUnit(unit)
    self.x = GetUnitX(unit)
    self.y = GetUnitY(unit)
    return self
end

---@param unit unit
function cls:UnitMoveTo(unit)
    SetUnitPosition(unit, self.x, self.y)
    return self
end

---@param other Vector2
function cls:SetTo(other)
    self.x = other.x
    self.y = other.y
    return
end

---@param other Vector2
function cls:Add(other)
    self.x = self.x + other.x
    self.y = self.y + other.y
    return self
end

---@param other Vector2
function cls:Sub(other)
    self.x = self.x - other.x
    self.y = self.y - other.y
    return self
end

---@param d real
function cls:Div(d)
    self.x = self.x / d
    self.y = self.y / d
    return self
end

---@param d real
function cls:Mul(d)
    self.x = self.x * d
    self.y = self.y * d
    return self
end

function cls:SetNormalize()
    local magnitude = self:Magnitude()

    if magnitude > 1e-05 then
        self.x = self.x / magnitude
        self.y = self.y / magnitude
    else
        self.x = 0
        self.y = 0
    end

    return self
end

function cls:Normalized()
    return self:Clone():SetNormalize()
end

function cls:SetMagnitude(len)
    self:SetNormalize():Mul(len)
    return self
end

---@param angle real radians
function cls:RotateSelf(angle)
    local cos = math.cos(angle)
    local sin = math.sin(angle)
    local x = cos * self.x - sin * self.y
    local y = sin * self.x + cos * self.y
    self.x = x
    self.y = y
    return self
end

---@param angle real radians
function cls:Rotate(angle)
    return self:Clone():RotateSelf(angle)
end

function cls:Clone()
    return new(self.x, self.y)
end

function cls:GetTerrainZ()
    MoveLocation(cls._loc, self.x, self.y)
    return GetLocationZ(cls._loc)
end

function cls:Magnitude()
    return m_sqrt(self.x * self.x + self.y * self.y)
end

function cls:MagnitudeSqr()
    return self.x * self.x + self.y * self.y
end

---@return string
function cls:tostring()
    return string.format("(%f,%f)", self.x, self.y)
end

function cls.__index(_, k)
    return rawget(cls, k)
end

function cls.__add(a, b)
    return new(a.x + b.x, a.y + b.y)
end

---@return Vector2
function cls.__sub(a, b)
    return new(a.x - b.x, a.y - b.y)
end

function cls.__div(v, d)
    return new(v.x / d, v.y / d)
end

function cls.__mul(a, d)
    if type(d) == "number" then
        return new(a.x * d, a.y * d)
    else
        return new(a * d.x, a * d.y)
    end
end

function cls.__unm(v)
    return new(-v.x, -v.y)
end

function cls.__eq(a, b)
    return ((a.x - b.x) ^ 2 + (a.y - b.y) ^ 2) < 9.999999e-11
end

setmetatable(cls, cls)

return cls

File: CSProject/Utils/Vector2.cs
using SFLib.Contracts;

#pragma warning disable CS0660 // Type defines operator == or operator != but does not override Object.Equals(object o)
#pragma warning disable CS0661 // Type defines operator == or operator != but does not override Object.GetHashCode()
public struct Vector2 : IEquatable<Vector2>
{
    public static Vector2 Zero => new(0, 0);

    private static location _loc = Location(0, 0);

    public static Vector2 InsideUnitCircle()
    {
        var angle = math.random() * 2 * math.pi;
        return new Vector2(math.cos(angle), math.sin(angle));
    }

    public static float Dot(Vector2 a, Vector2 b)
    {
        return a.x * b.x + a.y * b.y;
    }

    public static float Cross(Vector2 a, Vector2 b)
    {
        return a.y * b.x - a.x * b.y;
    }

    public static Vector2 operator -(Vector2 a)
    {
        return new Vector2(-a.x, -a.y);
    }

    public static Vector2 operator +(Vector2 a, Vector2 b)
    {
        return new Vector2(a.x + b.x, a.y + b.y);
    }

    public static Vector2 operator -(Vector2 a, Vector2 b)
    {
        return new Vector2(a.x - b.x, a.y - b.y);
    }

    public static Vector2 operator *(Vector2 v, float f)
    {
        return new Vector2(v.x * f, v.y * f);
    }

    public static Vector2 operator *(float f, Vector2 v)
    {
        return new Vector2(v.x * f, v.y * f);
    }

    public static Vector2 operator /(Vector2 v, float f)
    {
        return new Vector2(v.x / f, v.y / f);
    }

    public static bool operator ==(Vector2 a, Vector2 b)
    {
        return math.abs(a.x - b.x) < 0.0001f && math.abs(a.y - b.y) < 0.0001f;
    }

    public static bool operator !=(Vector2 a, Vector2 b)
    {
        return !(a == b);
    }

    public static float UnitDistance(unit a, unit b)
    {
        var v1 = FromUnit(a);
        var v2 = FromUnit(b);
        return (v1 - v2).Magnitude;
    }

    public static float SqrUnitDistance(unit a, unit b)
    {
        var v1 = FromUnit(a);
        var v2 = FromUnit(b);
        return (v1 - v2).SqrMagnitude;
    }

    public static Vector2 FromUnit(unit u)
    {
        return new Vector2(GetUnitX(u), GetUnitY(u));
    }

    public float x;
    public float y;

    public float Magnitude => math.sqrt(SqrMagnitude);
    public float SqrMagnitude => x * x + y * y;

    public Vector2 Normalized
    {
        get
        {
            var mag = Magnitude;
            if (mag < 0.0001f) return Zero;
            return this / mag;
        }
    }

    public Vector2(float x, float y)
    {
        this.x = x;
        this.y = y;
    }

    public Vector2 ClampMagnitude(float mag)
    {
        return Normalized * mag;
    }

    public bool Equals(Vector2 other)
    {
        return this == other;
    }

    public override string ToString()
    {
        return $"({x}, {y})";
    }

    public Vector2 Rotate(float angle)
    {
        var cos = math.cos(angle);
        var sin = math.sin(angle);
        return new Vector2(x * cos - y * sin, x * sin + y * cos);
    }

    public void UnitMoveTo(unit u)
    {
        SetUnitX(u, x);
        SetUnitY(u, y);
    }

    public float GetTerrainZ()
    {
        MoveLocation(_loc, x, y);
        return GetLocationZ(_loc);
    }
}

#pragma warning restore CS0661 // Type defines operator == or operator != but does not override Object.GetHashCode()
#pragma warning restore CS0660 // Type defines operator == or operator != but does not override Object.Equals(object o)